### PR TITLE
[wip] Circles support

### PR DESCRIPTION
**Pull request by tmcw**
Basic support to allow circle in v8 styles. Both the circle_bucket and
draw_circle need to be implemented.

This stub will not draw circles, but is sufficient to keep from failing
when it encounters a circle in a stylesheet.

* Previously: https://github.com/mapbox/mapbox-gl-js/pull/282

---

> if you use gl.POINTS you don't need to use indexed drawing. You don't need an element buffer and you would use drawArrays instead of drawElements. The tradeoffs:
> ansis: gl.POINTS only needs one vertex in the buffer per point. gl.TRIANGLES needs 4 vertices.
> ansis: gl.POINTS has a maximum size that varies depending on the platform. gl.TRIANGLES can be any size. I'm trying to find stats on what the minimum max size is for points.

The maximum size of a circle appears to be 64 https://developer.apple.com/opengl/capabilities/

With that limitation, we'll need to use triangles

> you'd draw the circular part in the shader. there would be two triangles to make a square. the fragment shader would figure out the distance of each pixel from the center and shade it appropriately

### CODE REVIEW

**Comment by tmcw** [comment]
I'm grabbing this branch and working on it today/tomorrow

**Comment by tmcw** [comment]
CIRCLES! This branch is ready for review: it draws circles of various colors, with blurring. I assume we'll merge this into the v8 branch if it passes muster?

/cc @kkaefer @ansis @jfirebaugh @mourner

File: shaders/circle.vertex.glsl [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+attribute vec2 a_extrude;
+
+uniform mat4 u_matrix;
+uniform mat4 u_exmatrix;
+
+varying vec2 v_extrude;
+
+void main(void) {
+    v_extrude = a_extrude;
+    vec4 extrude = u_exmatrix * vec4(a_extrude * u_size / 2.0 * sqrt(2.0), 0, 0);
```
Comment (ansis) [inline shaders/circle.vertex.glsl]: If you switch from a rotated square to an axis aligned square with vertices ((-1, -1), (1, -1), (1, 1), (-1, 1)) you should be able to remove the sqrt(2.0) here

File: js/data/circle_bucket.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        buffers.circleVertex,
+        buffers.circleElement);
+}
+
+CircleBucket.prototype.addFeatures = function() {
+    for (var i = 0; i < this.features.length; i++) {
+
+        this.elementGroups.makeRoomFor(6);
+
+        var geometry = this.features[i].loadGeometry()[0][0],
```
Comment (ansis) [inline js/data/circle_bucket.js]: should it add all vertices instead of just the first one?

**Comment by ansis** [comment]
We can reduce the size of circle vertex buffers by half if we pack the position and extrude together into 4 bytes. If we [switch to a square](https://github.com/mapbox/mapbox-gl-js/pull/1241#discussion_r31753712), there will be only two possible extrude component values: -1 and 1. Each component of the extrude only needs a single byte. Each position component has 16 bytes but doesn't need all of them.

The x position and x extrude could be combined into a single short with something like this:
`Math.round(x * 2) + Math.round(extrudeX / 2 + 0.5)`

They would need to be unpacked in the vertex shader.

We do something similar for lines: [buffer](https://github.com/mapbox/mapbox-gl-js/blob/master/js/data/buffer/line_vertex_buffer.js#L36), [vertex shader](https://github.com/mapbox/mapbox-gl-js/blob/a9d2c42e7b1e51f763f2ace825fc61867daaa4b9/shaders/line.vertex.glsl#L36).

**Comment by tmcw** [comment]
* Switched to axis-aligned boxes
* Packed the data into fewer bytes

Remaining issue

* It doesn't seem like pixel scaling is working correctly: a circle that should be 100px across is 134px, and the same if i drop out the shader magic and just draw the square that's behind it.

**Comment by tristen** [comment]
:eyes:

**Comment by ansis** [comment]
b2b74f0 fixes the size problem. `gl_Position`gets divided by `gl_Position.w` after the shader runs. The fix is to multiply the extrude by `.w` so that extrude isn't affected by the `.w` value.

**Comment by tmcw** [comment]
:+1: @ansis @jfirebaugh any remaining tasks before we should merge this into the v8 branch?

**Comment by ansis** [comment]
The circle needs to be antialiased. Something needs to be added [here](https://github.com/mapbox/mapbox-gl-js/blob/b2b74f06874f21983b07ee3436b346b12e2aef22/js/render/draw_circle.js#L22) so that 1px is always blurred.

Also, what are the units for `circle-blur`?

File: shaders/circle.vertex.glsl [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+uniform mat4 u_matrix;
+uniform mat4 u_exmatrix;
+
+varying vec2 v_extrude;
+
+void main(void) {
+    // unencode the extrusion vector that we snuck into the a_pos vector
+    v_extrude = vec2(
+        mod(a_pos.x, 2.0) * 2.0 - 1.0,
+        mod(a_pos.y, 2.0) * 2.0 - 1.0);
```
Comment (ansis) [inline shaders/circle.vertex.glsl]: Most glsl operations can work with vectors, so this can be written as `v_extrude = vec2(mod(a_pos, 2.0) * 2.0 - 1.0);`

**Comment by tmcw** [comment]
> The circle needs to be antialiased. Something needs to be added here so that 1px is always blurred.

Done

> Also, what are the units for circle-blur?

It's relative to the radius of the circle... blur: 1 is "fully blurred" afaik, where the gradient from center to outside is linear

**Comment by tmcw** [comment]
@ansis @jfirebaugh - I've added anti-aliasing and expanded the documentation of the blur parameter in the v8 style spec. Anything else remaining?

**Comment by peterqliu** [comment]
>It's relative to the radius of the circle... blur: 1 is "fully blurred" 

`line-blur` measures in [pixels](https://github.com/mapbox/mapbox-gl-style-spec/blob/mb-pages/docs/index.html#L1330). I'm not partial to either, but let's pick one

**Comment by tmcw** [comment]
I think we should not pick one: relative blur makes sense for circles and their use cases, and line-blur as pixels makes sense for lines. Let's keep this as-is.

**Comment by tmcw** [comment]
@kkaefer @ansis added retina/non-retina handling. I think this is good to merge. Final review?

**Comment by ansis** [comment]
Yes!

**Comment by tmcw** [comment]
Merged in 59c7af4ff337325e79a460e69f97d4151d2071da

**Comment by jfirebaugh** [comment]
\o/

**Comment by kkaefer** [comment]
\o/

**Pull request closed without merging**